src/Servers/Kestrel/Core/src/Internal/Http2/Http2OutputProducer.cs

```diff
@@ -429,16 +420,15 @@ private async Task ProcessDataWrites()
                         // Write any remaining content then write trailers
 
                         _stream.ResponseTrailers.SetReadOnly();
-                        _stream.DecrementActiveClientStreamCount();
```

Comment:
Rather than making any Http2FrameWriter changes, I think it would be simpler to move two previous calls to `DecrementActiveClientStreamCount` that occur before the completion of `WriteDataAsync` and `WriteDataAndTrailersAsync` until after they're awaited. All the other calls to `DecrementActiveClientStreamCount` could stay where they were before.